src/appleseed/renderer/modeling/object/proceduralobject.h

```diff
@@ -66,19 +67,28 @@ class APPLESEED_DLLSYMBOL ProceduralObject
     // Compute the intersection between a ray expressed in object space and
     // the surface of this object and return detailed intersection results.
     virtual void intersect(
-        const ShadingRay&       ray,
-        IntersectionResult&     result) const = 0;
+        const ShadingRay&               ray,
+        IntersectionResult&             result) const = 0;
 
     // Compute the intersection between a ray expressed in object space and
     // the surface of this object and simply return whether there was a hit.
     virtual bool intersect(
-        const ShadingRay&       ray) const = 0;
+        const ShadingRay&               ray) const = 0;
+
+    // Compute a front point, a back point and the geometric normal in object
+    // instance space for a given ray with origin being a point on the surface
+    // of the object.
+    virtual void refine_and_offset(
+        const foundation::Ray3d&        obj_inst_ray,
```

Comment:
I can't because transforming a ShadingRay returns a Ray
![image](https://user-images.githubusercontent.com/4656466/67636342-f68f0e80-f8cf-11e9-957a-e9f3165a04ae.png)